# CalCOFI-Investigaciones Cooperativas de Pesca Oceánica de California
Datos oceanográficos

-Un poco de historia

Este conjunto comenzó en el año 1949 y continúa en el presente recopilando una serie de datos hidrográficos, físicos, químicos y biológicos tanto en lugares delimitados como durante el trayecto. Recolectados a intervalos regulares de tiempo y espacio se convirtieron rápidamente en valiosos para documentar los ciclos climáticos en la Corriente de California y diversas respuestas biológicas a éstos. La investigación de CalCOFI atrajo la atención mundial sobre la respuesta biológica al drástico calentamiento del Pacífico en 1957-58 e introdujo el término "El Niño" en la literatura científica. 



# Dentro

• Con más de 50 000 estaciones de muestreo, se han lanzado al mar botellas a profundidades de hasta 500 m que recolectaron variables como: 

Temperatura, salinidad, oxígeno, fosfato, silicato, nitrato y nitrito, clorofila, transmitancia, PAR, productividad primaria C14, biodiversidad del fitoplancton, biomasa del zooplancton y biodiversidad del zooplancton.

• Incluye información sobre la abundancia de larvas de más de 250 especies de peces y huevos de especies comerciales clave

Tenemos por lo tanto 3 Dataset con información:

- Bottle: el estudio con botella Niskin utiliza este cilindro especial de PVC para capturar muestras de agua a diferentes profundidades sin contaminarlas
- Cast:  en qué lugar ocurrieron estos lanzamientos, hora fecha y otras condiciones climatológicas
- Biologico:  Sobre biodiversidad marina en esos lugares

# Bottle

In [1]:
import pandas as pd

# trabajando en jupyter SUBIR UN NIVEL DE CARPETA (../)
bottle = pd.read_csv('../data/raw/bottle.csv', low_memory=False)

# Obtener las dimensiones
print(f"Bottle: Filas: {bottle.shape[0]}, Columnas: {bottle.shape[1]}")

Bottle: Filas: 864863, Columnas: 74


In [2]:

# Ver primeras lineas de Bottle
bottle.head()

,Cst_Cnt,Btl_Cnt,Sta_ID,Depth_ID,Depthm,T_degC,Salnty,O2ml_L,STheta,O2Sat,...,R_PHAEO,R_PRES,R_SAMP,DIC1,DIC2,TA1,TA2,pH2,pH1,DIC Quality Comment
0,1,1,054.0 056.0,19-4903CR-HY-060-0930-05400560-0000A-3,0,10.50,33.440,NaN,25.649,NaN,...,NaN,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,1,2,054.0 056.0,19-4903CR-HY-060-0930-05400560-0008A-3,8,10.46,33.440,NaN,25.656,NaN,...,NaN,8,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,1,3,054.0 056.0,19-4903CR-HY-060-0930-05400560-0010A-7,10,10.46,33.437,NaN,25.654,NaN,...,NaN,10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,1,4,054.0 056.0,19-4903CR-HY-060-0930-05400560-0019A-3,19,10.45,33.420,NaN,25.643,NaN,...,NaN,19,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,1,5,054.0 056.0,19-4903CR-HY-060-0930-05400560-0020A-7,20,10.45,33.421,NaN,25.643,NaN,...,NaN,20,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [3]:

# Obtener información sobre tipos de datos y valores no nulos

bottle.info()

<class 'pandas.DataFrame'>
RangeIndex: 864863 entries, 0 to 864862
Data columns (total 74 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   Cst_Cnt              864863 non-null  int64  
 1   Btl_Cnt              864863 non-null  int64  
 2   Sta_ID               864863 non-null  str    
 3   Depth_ID             864863 non-null  str    
 4   Depthm               864863 non-null  int64  
 5   T_degC               853900 non-null  float64
 6   Salnty               817509 non-null  float64
 7   O2ml_L               696201 non-null  float64
 8   STheta               812174 non-null  float64
 9   O2Sat                661274 non-null  float64
 10  Oxy_µmol/Kg          661268 non-null  float64
 11  BtlNum               118667 non-null  float64
 12  RecInd               864863 non-null  int64  
 13  T_prec               853900 non-null  float64
 14  T_qual               23127 non-null   float64
 15  S_prec               817509 

# Conclusión
Bottle nos habla sobre el método cientifico de las botellas de Niskin, dispuesta en forma de roseta, recopilaron datos físicos y químicos que dieron lugar a otro estudio cientifico llamado isótopo Carbono-14 (¹⁴C) Es la velocidad o tasa con la que el fitoplancton transforman el  CO₂ en materia orgánica

De este dataset solo vamos a centrarnos en los datos generales, no en el estudio del C14

# Cast

In [4]:
cast = pd.read_csv('../data/raw/cast.csv', low_memory=False)

# Obtener las dimensiones
print(f" Cast: Filas: {cast.shape[0]}, Columnas: {cast.shape[1]}")

 Cast: Filas: 34404, Columnas: 61


In [5]:
# Ver primeras lineas de Head

cast.head()

,Cst_Cnt,Cruise_ID,Cruise,Cruz_Sta,DbSta_ID,Cast_ID,Sta_ID,Quarter,Sta_Code,Distance,...,Wave_Prd,Wind_Dir,Wind_Spd,Barometer,Dry_T,Wet_T,Wea,Cloud_Typ,Cloud_Amt,Visibility
0,1,1949-03-01-C-31CR,194903,19490305400560,5400560,19-4903CR-HY-060-0930-05400560,054.0 056.0,1,NST,NaN,...,NaN,23.0,18.0,NaN,NaN,NaN,2.0,NaN,NaN,NaN
1,2,1949-03-01-C-31CR,194903,19490305200750,5200750,19-4903CR-HY-060-2112-05200750,052.0 075.0,1,NST,NaN,...,NaN,16.0,5.0,NaN,NaN,NaN,4.0,NaN,NaN,NaN
2,3,1949-03-01-C-31CR,194903,19490305100850,5100850,19-4903CR-HY-061-0354-05100850,051.0 085.0,1,NST,NaN,...,NaN,23.0,5.0,NaN,NaN,NaN,6.0,NaN,NaN,NaN
3,4,1949-03-01-C-31CR,194903,19490305000950,5000950,19-4903CR-HY-061-1042-05000950,050.0 095.0,1,NST,NaN,...,NaN,18.0,8.0,NaN,NaN,NaN,2.0,NaN,NaN,NaN
4,5,1949-03-01-C-31CR,194903,19490305001040,5001040,19-4903CR-HY-061-1706-05001040,050.0 104.0,1,NST,NaN,...,NaN,27.0,13.0,NaN,NaN,NaN,7.0,NaN,NaN,NaN


In [6]:

# Obtener información sobre tipos de datos y valores no nulos de Cast
cast.info()

<class 'pandas.DataFrame'>
RangeIndex: 34404 entries, 0 to 34403
Data columns (total 61 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Cst_Cnt      34404 non-null  int64  
 1   Cruise_ID    34404 non-null  str    
 2   Cruise       34404 non-null  int64  
 3   Cruz_Sta     34404 non-null  int64  
 4   DbSta_ID     34404 non-null  int64  
 5   Cast_ID      34404 non-null  str    
 6   Sta_ID       34404 non-null  str    
 7   Quarter      34404 non-null  int64  
 8   Sta_Code     34404 non-null  str    
 9   Distance     19714 non-null  float64
 10  Date         34404 non-null  str    
 11  Year         34404 non-null  int64  
 12  Month        34404 non-null  int64  
 13  Julian_Date  34404 non-null  int64  
 14  Julian_Day   34404 non-null  int64  
 15  Time         34080 non-null  str    
 16  Lat_Dec      34404 non-null  float64
 17  Lat_Deg      34404 non-null  int64  
 18  Lat_Min      34404 non-null  float64
 19  Lat_Hem      34

# Conclusión
Cast nos habla sobre las "lanzadas" de botellas al mar, dónde, cuándo, a qué profundidad fueron etc 

# Biológico

Dentro de Biologico hay más información y Datasets. Vamos a ver que hay dentro de cada uno para ver cuál nos sirve en nuestro análisis


In [7]:
import pandas as pd

eml = pd.read_xml ("../data/raw/biologico/eml.xml")
eml.head()



,alternateIdentifier,shortName,title,creator,pubDate,language,abstract,keywordSet,intellectualRights,licensed,distribution,coverage,maintenance,contact,methods,project,metadata
0,https://ipt-obis.gbif.us/resource?r=calcofi_ic...,calcofi_ichthyo,CalCOFI Fish Larvae & Egg Tows,\n,\n 2026-03-27\n,eng,\n Fish larvae counts and standardi...,\n,\n,\n,\n,\n,\n,\n,\n,\n,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,\n


In [8]:
# Obtener información sobre tipos de datos y valores no nulos de Eml

eml.info()



<class 'pandas.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   alternateIdentifier  1 non-null      str  
 1   shortName            1 non-null      str  
 2   title                1 non-null      str  
 3   creator              1 non-null      str  
 4   pubDate              1 non-null      str  
 5   language             1 non-null      str  
 6   abstract             1 non-null      str  
 7   keywordSet           1 non-null      str  
 8   intellectualRights   1 non-null      str  
 9   licensed             1 non-null      str  
 10  distribution         1 non-null      str  
 11  coverage             1 non-null      str  
 12  maintenance          1 non-null      str  
 13  contact              1 non-null      str  
 14  methods              1 non-null      str  
 15  project              1 non-null      str  
 16  metadata             1 non-null      str 

In [9]:
# Obtener las dimensiones de Eml
print(f"Eml: Filas: {eml.shape[0]}, Columnas: {eml.shape[1]}")

Eml: Filas: 2, Columnas: 17


# Conclusión 
El dataset EML no nos arroja datos necesarios para nuestro estudio porque habla de libros

# Meta

In [10]:
# Obtener las dimensiones de Meta
meta = pd.read_xml ("../data/raw/biologico/meta.xml")
meta.head()



,encoding,fieldsTerminatedBy,linesTerminatedBy,fieldsEnclosedBy,ignoreHeaderLines,rowType,files,id,field,coreid
0,UTF-8,\t,\n,NaN,1,http://rs.tdwg.org/dwc/terms/Event,\n,NaN,NaN,NaN
1,UTF-8,\t,\n,NaN,1,http://rs.iobis.org/obis/terms/ExtendedMeasure...,\n,NaN,NaN,NaN
2,UTF-8,\t,\n,NaN,1,http://rs.tdwg.org/dwc/terms/Occurrence,\n,NaN,NaN,NaN


In [11]:
# Obtener información sobre tipos de datos y valores no nulos de Meta

meta.info()


<class 'pandas.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   encoding            3 non-null      str    
 1   fieldsTerminatedBy  3 non-null      str    
 2   linesTerminatedBy   3 non-null      str    
 3   fieldsEnclosedBy    0 non-null      float64
 4   ignoreHeaderLines   3 non-null      int64  
 5   rowType             3 non-null      str    
 6   files               3 non-null      str    
 7   id                  0 non-null      float64
 8   field               0 non-null      float64
 9   coreid              0 non-null      float64
dtypes: float64(4), int64(1), str(5)
memory usage: 372.0 bytes


In [12]:
# Obtener las dimensiones de Meta
print(f"Meta: Filas: {meta.shape[0]}, Columnas: {meta.shape[1]}")

Meta: Filas: 3, Columnas: 10


# Conclusión 
Meta puede llegar a interesarnos porque contiene los nombres científicos y específicos de cada especie para identificarlas

# Event

In [13]:

event = pd.read_csv('../data/raw/biologico/event.txt', sep=None, engine='python')

# Muestra las primeras filas del dataset de Event
event.head()

,id,datasetID,eventID,parentEventID,eventType,eventDate,habitat,samplingProtocol,sampleSizeValue,sampleSizeUnit,eventRemarks,locationID,decimalLatitude,decimalLongitude,geodeticDatum,coordinateUncertaintyInMeters,footprintWKT
0,510147,calcofi.io_workflows_ichthyo_to_obis_2026-03-06,510147,NaN,cruise,1951-01-19T07:25:00Z/1951-01-29T19:26:00Z,marine,Marine plankton survey cruise,NaN,NaN,Cruise on SCOFIELD from 1951-01-19 to 1951-01-29,NaN,NaN,NaN,WGS84,1000,"POLYGON((-124.5 31.983333, -118.625 31.983333,..."
1,5101CR,calcofi.io_workflows_ichthyo_to_obis_2026-03-06,5101CR,NaN,cruise,1951-01-09T18:16:00Z/1951-01-23T03:06:00Z,marine,Marine plankton survey cruise,NaN,NaN,Cruise on CREST from 1951-01-09 to 1951-01-23,NaN,NaN,NaN,WGS84,1000,"POLYGON((-124.073333 28.675, -116.583333 28.67..."
2,5101HO,calcofi.io_workflows_ichthyo_to_obis_2026-03-06,5101HO,NaN,cruise,1951-01-10T03:40:00Z/1951-01-24T10:06:00Z,marine,Marine plankton survey cruise,NaN,NaN,Cruise on HORIZON from 1951-01-10 to 1951-01-24,NaN,NaN,NaN,WGS84,1000,"POLYGON((-121.096666 24.323333, -113.113333 24..."
3,5102BD,calcofi.io_workflows_ichthyo_to_obis_2026-03-06,5102BD,NaN,cruise,1951-02-09T07:40:00Z/1951-02-25T08:55:00Z,marine,Marine plankton survey cruise,NaN,NaN,Cruise on BLACK DOUGLAS from 1951-02-09 to 195...,NaN,NaN,NaN,WGS84,1000,"POLYGON((-125.933333 28.825, -116.9 28.825, -1..."
4,5102CR,calcofi.io_workflows_ichthyo_to_obis_2026-03-06,5102CR,NaN,cruise,1951-02-06T20:00:00Z/1951-02-23T13:26:00Z,marine,Marine plankton survey cruise,NaN,NaN,Cruise on CREST from 1951-02-06 to 1951-02-23,NaN,NaN,NaN,WGS84,1000,"POLYGON((-120.233333 24.333333, -113.075 24.33..."


In [14]:
# Obtener información sobre tipos de datos y valores no nulos de Event

event.info()

<class 'pandas.DataFrame'>
RangeIndex: 77188 entries, 0 to 77187
Data columns (total 17 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   id                             77188 non-null  str    
 1   datasetID                      77188 non-null  str    
 2   eventID                        77188 non-null  str    
 3   parentEventID                  76512 non-null  str    
 4   eventType                      77188 non-null  str    
 5   eventDate                      77188 non-null  str    
 6   habitat                        676 non-null    str    
 7   samplingProtocol               77188 non-null  str    
 8   sampleSizeValue                76512 non-null  float64
 9   sampleSizeUnit                 76512 non-null  str    
 10  eventRemarks                   77188 non-null  str    
 11  locationID                     76512 non-null  str    
 12  decimalLatitude                76512 non-null  float64
 1

In [15]:
# Obtener las dimensiones de Event
print(f"Event: Filas: {event.shape[0]}, Columnas: {event.shape[1]}")

Event: Filas: 77188, Columnas: 17


# Conclusión

Event en principio puede llegar a interesarnos porque habla de  dónde y qué especies que se encontraron en cada viaje, especificando los barcos y a qué profundidad  

In [16]:
extended = pd.read_csv('../data/raw/biologico/extendedmeasurementorfact.txt', sep=None, engine='python')

# Muestra las primeras filas del dataset de Extended
extended.head()


,id,measurementID,occurrenceID,measurementType,measurementTypeID,measurementValue,measurementUnit,measurementUnitID,measurementMethod,measurementRemarks
0,df8733f2-02eb-ef11-a09e-2cea7fa0979c,7b2dcb22-97be-5b70-839d-09ec3ff4c164,NaN,standardized haul factor,NaN,4.4000,dimensionless,http://vocab.nerc.ac.uk/collection/P06/current...,https://oceanview.pfeg.noaa.gov/CalCOFI/calcof...,Standardization factor for water filtered; abu...
1,df8733f2-02eb-ef11-a09e-2cea7fa0979c,73357e2f-d5c3-5def-beb7-583855e6f8c4,NaN,proportion of sample sorted,NaN,1.0000,dimensionless,http://vocab.nerc.ac.uk/collection/P06/current...,https://oceanview.pfeg.noaa.gov/CalCOFI/calcof...,Fraction of total sample examined
2,df8733f2-02eb-ef11-a09e-2cea7fa0979c,1636da6e-cd4f-5b2d-a110-8e23d20471c0,NaN,small plankton biomass,NaN,6.0125,grams,http://vocab.nerc.ac.uk/collection/P06/current...,https://oceanview.pfeg.noaa.gov/CalCOFI/calcof...,NaN
3,df8733f2-02eb-ef11-a09e-2cea7fa0979c,0080735d-a045-56c2-906d-6ef0a19953f3,NaN,total plankton biomass,NaN,6.0125,grams,http://vocab.nerc.ac.uk/collection/P06/current...,https://oceanview.pfeg.noaa.gov/CalCOFI/calcof...,NaN
4,e08733f2-02eb-ef11-a09e-2cea7fa0979c,2a59012a-79ac-547f-bed3-ba6d7757851b,NaN,standardized haul factor,NaN,0.0200,dimensionless,http://vocab.nerc.ac.uk/collection/P06/current...,https://oceanview.pfeg.noaa.gov/CalCOFI/calcof...,Standardization factor for water filtered; abu...


In [17]:
# Obtener información sobre tipos de datos y valores no nulos de Extended

extended.info()

<class 'pandas.DataFrame'>
RangeIndex: 610816 entries, 0 to 610815
Data columns (total 10 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   id                  610816 non-null  str    
 1   measurementID       610816 non-null  str    
 2   occurrenceID        367218 non-null  str    
 3   measurementType     610816 non-null  str    
 4   measurementTypeID   0 non-null       float64
 5   measurementValue    610816 non-null  float64
 6   measurementUnit     610816 non-null  str    
 7   measurementUnitID   610816 non-null  str    
 8   measurementMethod   610816 non-null  str    
 9   measurementRemarks  519452 non-null  str    
dtypes: float64(2), str(8)
memory usage: 46.6 MB


# Conclusión 
Más medidas complementarias de Event. En principio no nos interesa Extended porque habla de cómo se tomaron esas muestras, cómo fueron transladadas, cuánto se tomó de ellas, cosas que no nos parecen relevantes para nuestro análisis

# Ocurrence

In [18]:
occurrence = pd.read_csv('../data/raw/biologico/occurrence.txt', sep=None, engine='python')

# Primeras filas del dataset de Occurrence
occurrence.head()

,id,modified,basisOfRecord,occurrenceID,organismQuantity,organismQuantityType,lifeStage,occurrenceStatus,preparations,eventID,scientificNameID,scientificName,kingdom
0,eb8733f2-02eb-ef11-a09e-2cea7fa0979c,2026-03-06,HumanObservation,d7d7f5c3-17d2-549c-9470-0008891108e2,9,individuals,egg,present,egg sample,eb8733f2-02eb-ef11-a09e-2cea7fa0979c,urn:lsid:marinespecies.org:taxname:293496,Teleostei,Animalia
1,eb8733f2-02eb-ef11-a09e-2cea7fa0979c,2026-03-06,HumanObservation,dfa978d9-2788-5cdb-9dba-aa4755bb021c,1,individuals,egg,present,egg sample,eb8733f2-02eb-ef11-a09e-2cea7fa0979c,urn:lsid:marinespecies.org:taxname:272458,Merluccius productus,Animalia
2,ec8733f2-02eb-ef11-a09e-2cea7fa0979c,2026-03-06,HumanObservation,0360608f-cf81-5853-be4b-80cca27a1b23,20,individuals,egg,present,egg sample,ec8733f2-02eb-ef11-a09e-2cea7fa0979c,urn:lsid:marinespecies.org:taxname:293496,Teleostei,Animalia
3,ef8733f2-02eb-ef11-a09e-2cea7fa0979c,2026-03-06,HumanObservation,b08d9ed2-407c-56a4-b4d7-57d1b716f2af,70,individuals,egg,present,egg sample,ef8733f2-02eb-ef11-a09e-2cea7fa0979c,urn:lsid:marinespecies.org:taxname:293496,Teleostei,Animalia
4,f38733f2-02eb-ef11-a09e-2cea7fa0979c,2026-03-06,HumanObservation,4264c438-9c45-5303-a4f7-c916b109eb72,43,individuals,egg,present,egg sample,f38733f2-02eb-ef11-a09e-2cea7fa0979c,urn:lsid:marinespecies.org:taxname:293496,Teleostei,Animalia


In [19]:
# Obtener información sobre tipos de datos y valores no nulos de Occurrence

occurrence.info()

<class 'pandas.DataFrame'>
RangeIndex: 463655 entries, 0 to 463654
Data columns (total 13 columns):
 #   Column                Non-Null Count   Dtype
---  ------                --------------   -----
 0   id                    463655 non-null  str  
 1   modified              463655 non-null  str  
 2   basisOfRecord         463655 non-null  str  
 3   occurrenceID          463655 non-null  str  
 4   organismQuantity      463655 non-null  int64
 5   organismQuantityType  463655 non-null  str  
 6   lifeStage             463655 non-null  str  
 7   occurrenceStatus      463655 non-null  str  
 8   preparations          463655 non-null  str  
 9   eventID               463655 non-null  str  
 10  scientificNameID      463655 non-null  str  
 11  scientificName        463655 non-null  str  
 12  kingdom               463655 non-null  str  
dtypes: int64(1), str(12)
memory usage: 46.0 MB


# Conclusión

Ocurrence nos interesa porque habla de las especies encontradas, su nombre específico, en qué cantidad etc

# Conclusión de Biológico

• Descartamos:
Extended, 
Eml

• Tal vez más adelante: 
Event, 
Meta

• Aporta: 
Ocurrence

# Limpiando los datos de Bottle y Cast

In [3]:
# Eliminamos datos nulos de bottle para ver cuántos nulos hay por columna

nulos_bottle= bottle.isnull().sum()
print (nulos_bottle)

Cst_Cnt                     0
Btl_Cnt                     0
Sta_ID                      0
Depth_ID                    0
Depthm                      0
                        ...  
TA1                    862779
TA2                    864629
pH2                    864853
pH1                    864779
DIC Quality Comment    864808
Length: 74, dtype: int64


Tenemos muchos con mas de 862779 columnas sin datos


In [4]:

# conservamos columnas con menos del 80% de nulos
limite_nulos = len(bottle) * 0.80

# Conservamos solo columnas con pocos nulos
bottle_limpio = bottle.dropna(thresh=limite_nulos, axis=1)

print(f"Columnas originales: {bottle.shape[1]}")
print(f"Columnas útiles conservadas: {bottle_limpio.shape[1]}")


Columnas originales: 74
Columnas útiles conservadas: 26


In [22]:
# Hacemos lo mismo con Cast

nulos_cast= cast.isnull().sum()
print (nulos_cast)

Cst_Cnt           0
Cruise_ID         0
Cruise            0
Cruz_Sta          0
DbSta_ID          0
              ...  
Wet_T         15415
Wea            7597
Cloud_Typ     24484
Cloud_Amt     22522
Visibility    24897
Length: 61, dtype: int64


Tenemos muchos con mas de 24897 columnas sin datos de Cast

In [23]:
# Calcular el porcentaje de nulos por columna de Cast

import pandas as pd
import sqlite3


porcentaje_nulos = cast.isnull().mean()

#  Conservar los nombres de las columnas con MENOS del 80% de nulos
columnas_utiles = porcentaje_nulos[porcentaje_nulos < 0.80].index


In [24]:

# Filtrar el DataFrame seleccionando directamente esas columnas
cast_limpio = cast[columnas_utiles]

print(f"Columnas originales: {cast.shape[1]}")
print(f"Columnas útiles conservadas: {cast_limpio.shape[1]}")

Columnas originales: 61
Columnas útiles conservadas: 52


In [25]:
import pandas as pd

bottle = pd.read_csv('../data/raw/bottle.csv', low_memory=False)

# Eliminar valores duplicados
doble_bottle = bottle.duplicated().sum()

print(f"Total de filas duplicadas exactas de Bottle: {doble_bottle}")

Total de filas duplicadas exactas de Bottle: 0


In [26]:
cast = pd.read_csv('../data/raw/cast.csv', low_memory=False)

doble_cast = cast.duplicated().sum()

print(f"Total de filas duplicadas exactas de Cast: {doble_cast}")

Total de filas duplicadas exactas de Cast: 0


No hay filas duplicadas ni en Bottle ni en Cast

# Pasando todo a base de datos sqlite


In [27]:
# Pasando todo a base de datos sqlite
# 1ro Bottle

import sqlite3
import pandas as pd

botellas= "../data/raw/bottle.csv"
bottle = "bottle.db"
las_botellas = "bottle"

#  Leer el archivo CSV con Pandas
bottle = pd.read_csv(botellas)

#  Conectar a SQLite (Crea el archivo .db si no existe)
conexion = sqlite3.connect("bottle.db")

# Guardar el DataFrame en la base de datos como una tabla
bottle.to_sql(las_botellas, conexion, if_exists="replace", index=False)


print(f"¡Base de datos '{bottle}' creada con éxito con la tabla '{las_botellas}'!")

/tmp/ipykernel_64834/3166508059.py:12: DtypeWarning: Columns (0: IncTim, 1: DIC Quality Comment) have mixed types. Specify dtype option on import or set low_memory=False.
  bottle = pd.read_csv(botellas)


¡Base de datos '        Cst_Cnt  Btl_Cnt       Sta_ID                                Depth_ID  \
0             1        1  054.0 056.0  19-4903CR-HY-060-0930-05400560-0000A-3   
1             1        2  054.0 056.0  19-4903CR-HY-060-0930-05400560-0008A-3   
2             1        3  054.0 056.0  19-4903CR-HY-060-0930-05400560-0010A-7   
3             1        4  054.0 056.0  19-4903CR-HY-060-0930-05400560-0019A-3   
4             1        5  054.0 056.0  19-4903CR-HY-060-0930-05400560-0020A-7   
...         ...      ...          ...                                     ...   
864858    34404   864859  093.4 026.4  20-1611SR-MX-310-2239-09340264-0000A-7   
864859    34404   864860  093.4 026.4  20-1611SR-MX-310-2239-09340264-0002A-3   
864860    34404   864861  093.4 026.4  20-1611SR-MX-310-2239-09340264-0005A-3   
864861    34404   864862  093.4 026.4  20-1611SR-MX-310-2239-09340264-0010A-3   
864862    34404   864863  093.4 026.4  20-1611SR-MX-310-2239-09340264-0015A-3   

        Dep

In [28]:
# Pasando todo a base de datos sqlite
# 2do Cast

import sqlite3
import pandas as pd

muestras = "../data/raw/cast.csv"
cast = "cast.db"
las_muestras = "cast"

#  Leer el archivo CSV con Pandas
cast = pd.read_csv(muestras)

#  Conectar a SQLite (Crea el archivo .db si no existe)
conexion = sqlite3.connect("cast.db")

# Guardar el DataFrame en la base de datos como una tabla
cast.to_sql(las_muestras, conexion, if_exists="replace", index=False)


print(f"¡Base de datos '{cast}' creada con éxito con la tabla '{las_muestras}'!")


/tmp/ipykernel_64834/813737589.py:12: DtypeWarning: Columns (0: Data_Or, 1: Cruz_Num, 2: Inc_Str, 3: Inc_End, 4: PST_LAN, 5: Civil_T) have mixed types. Specify dtype option on import or set low_memory=False.
  cast = pd.read_csv(muestras)


¡Base de datos '       Cst_Cnt          Cruise_ID  Cruise        Cruz_Sta  DbSta_ID  \
0            1  1949-03-01-C-31CR  194903  19490305400560   5400560   
1            2  1949-03-01-C-31CR  194903  19490305200750   5200750   
2            3  1949-03-01-C-31CR  194903  19490305100850   5100850   
3            4  1949-03-01-C-31CR  194903  19490305000950   5000950   
4            5  1949-03-01-C-31CR  194903  19490305001040   5001040   
...        ...                ...     ...             ...       ...   
34399    34400  2016-11-06-C-33SR  201611  20161109330900   9330900   
34400    34401  2016-11-06-C-33SR  201611  20161109331000   9331000   
34401    34402  2016-11-06-C-33SR  201611  20161109331100   9331100   
34402    34403  2016-11-06-C-33SR  201611  20161109331200   9331200   
34403    34404  2016-11-06-C-33SR  201611  20161109340264   9340264   

                               Cast_ID       Sta_ID  Quarter Sta_Code  \
0       19-4903CR-HY-060-0930-05400560  054.0 056.0       

# Exportar la base de datos SQLite a un archivo .sql completo

In [38]:
#  Exportar la base de datos SQLite a un archivo .sql completo
#  1. Bottle
conexion = sqlite3.connect("bottle.db")

with open(las_botellas, "w", encoding="utf-8") as f:
    for linea in conexion.iterdump():
        f.write(f"{linea}\n")

# 5. Cerrar la conexión
conexion.close()
print(f"¡Archivo SQL listo y guardado como '{bottle}'!")

¡Archivo SQL listo y guardado como '        Cst_Cnt  Btl_Cnt       Sta_ID                                Depth_ID  \
0             1        1  054.0 056.0  19-4903CR-HY-060-0930-05400560-0000A-3   
1             1        2  054.0 056.0  19-4903CR-HY-060-0930-05400560-0008A-3   
2             1        3  054.0 056.0  19-4903CR-HY-060-0930-05400560-0010A-7   
3             1        4  054.0 056.0  19-4903CR-HY-060-0930-05400560-0019A-3   
4             1        5  054.0 056.0  19-4903CR-HY-060-0930-05400560-0020A-7   
...         ...      ...          ...                                     ...   
864858    34404   864859  093.4 026.4  20-1611SR-MX-310-2239-09340264-0000A-7   
864859    34404   864860  093.4 026.4  20-1611SR-MX-310-2239-09340264-0002A-3   
864860    34404   864861  093.4 026.4  20-1611SR-MX-310-2239-09340264-0005A-3   
864861    34404   864862  093.4 026.4  20-1611SR-MX-310-2239-09340264-0010A-3   
864862    34404   864863  093.4 026.4  20-1611SR-MX-310-2239-09340264-001

In [39]:
#  Exportar la base de datos SQLite a un archivo .sql completo
#  2. Cast
conexion = sqlite3.connect("cast.db")
with open(las_muestras, "w", encoding="utf-8") as f:
    for linea in conexion.iterdump():
        f.write(f"{linea}\n")
        
# 5. Cerrar la conexión
conexion.close()

print(f"¡Archivo SQL listo y guardado como '{cast}'!")

¡Archivo SQL listo y guardado como '       Cst_Cnt          Cruise_ID  Cruise        Cruz_Sta  DbSta_ID  \
0            1  1949-03-01-C-31CR  194903  19490305400560   5400560   
1            2  1949-03-01-C-31CR  194903  19490305200750   5200750   
2            3  1949-03-01-C-31CR  194903  19490305100850   5100850   
3            4  1949-03-01-C-31CR  194903  19490305000950   5000950   
4            5  1949-03-01-C-31CR  194903  19490305001040   5001040   
...        ...                ...     ...             ...       ...   
34399    34400  2016-11-06-C-33SR  201611  20161109330900   9330900   
34400    34401  2016-11-06-C-33SR  201611  20161109331000   9331000   
34401    34402  2016-11-06-C-33SR  201611  20161109331100   9331100   
34402    34403  2016-11-06-C-33SR  201611  20161109331200   9331200   
34403    34404  2016-11-06-C-33SR  201611  20161109340264   9340264   

                               Cast_ID       Sta_ID  Quarter Sta_Code  \
0       19-4903CR-HY-060-0930-05400560

In [40]:
import sqlite3
import pandas as pd

conexion = sqlite3.connect('bottle.db')

# 1. Obtener la lista de columnas de la tabla
cols_bottle = pd.read_sql_query("PRAGMA table_info(bottle);", conexion)['name'].tolist()

conexion.close()

print(f" Columnas en 'bottle' ({len(cols_bottle)}):")
print(cols_bottle)

 Columnas en 'bottle' (74):
['Cst_Cnt', 'Btl_Cnt', 'Sta_ID', 'Depth_ID', 'Depthm', 'T_degC', 'Salnty', 'O2ml_L', 'STheta', 'O2Sat', 'Oxy_µmol/Kg', 'BtlNum', 'RecInd', 'T_prec', 'T_qual', 'S_prec', 'S_qual', 'P_qual', 'O_qual', 'SThtaq', 'O2Satq', 'ChlorA', 'Chlqua', 'Phaeop', 'Phaqua', 'PO4uM', 'PO4q', 'SiO3uM', 'SiO3qu', 'NO2uM', 'NO2q', 'NO3uM', 'NO3q', 'NH3uM', 'NH3q', 'C14As1', 'C14A1p', 'C14A1q', 'C14As2', 'C14A2p', 'C14A2q', 'DarkAs', 'DarkAp', 'DarkAq', 'MeanAs', 'MeanAp', 'MeanAq', 'IncTim', 'LightP', 'R_Depth', 'R_TEMP', 'R_POTEMP', 'R_SALINITY', 'R_SIGMA', 'R_SVA', 'R_DYNHT', 'R_O2', 'R_O2Sat', 'R_SIO3', 'R_PO4', 'R_NO3', 'R_NO2', 'R_NH4', 'R_CHLA', 'R_PHAEO', 'R_PRES', 'R_SAMP', 'DIC1', 'DIC2', 'TA1', 'TA2', 'pH2', 'pH1', 'DIC Quality Comment']


In [41]:
conexion = sqlite3.connect('cast.db')
cols_cast = pd.read_sql_query("PRAGMA table_info(cast);", conexion)['name'].tolist()

print(f"\nColumnas en 'cast' ({len(cols_cast)}):")
print(cols_cast)


Columnas en 'cast' (61):
['Cst_Cnt', 'Cruise_ID', 'Cruise', 'Cruz_Sta', 'DbSta_ID', 'Cast_ID', 'Sta_ID', 'Quarter', 'Sta_Code', 'Distance', 'Date', 'Year', 'Month', 'Julian_Date', 'Julian_Day', 'Time', 'Lat_Dec', 'Lat_Deg', 'Lat_Min', 'Lat_Hem', 'Lon_Dec', 'Lon_Deg', 'Lon_Min', 'Lon_Hem', 'Rpt_Line', 'St_Line', 'Ac_Line', 'Rpt_Sta', 'St_Station', 'Ac_Sta', 'Bottom_D', 'Secchi', 'ForelU', 'Ship_Name', 'Ship_Code', 'Data_Type', 'Order_Occ', 'Event_Num', 'Cruz_Leg', 'Orig_Sta_ID', 'Data_Or', 'Cruz_Num', 'IntChl', 'IntC14', 'Inc_Str', 'Inc_End', 'PST_LAN', 'Civil_T', 'TimeZone', 'Wave_Dir', 'Wave_Ht', 'Wave_Prd', 'Wind_Dir', 'Wind_Spd', 'Barometer', 'Dry_T', 'Wet_T', 'Wea', 'Cloud_Typ', 'Cloud_Amt', 'Visibility']


Seguimos teniendo los archivos originales, hay que sobreesribirlos

In [5]:
import sqlite3
import pandas as pd

# 1. Leer el CSV original (o cargar la tabla desde SQLite)
bottle = pd.read_csv("../data/raw/bottle.csv", low_memory=False)

# 2. Filtrar columnas con MENOS del 80% de nulos
porcentaje_nulos = bottle.isnull().mean()
cols_utiles_bottle = porcentaje_nulos[porcentaje_nulos < 0.80].index
df_bottle_limpio = bottle[cols_utiles_bottle]

# 3. Sobrescribir la base de datos bottle.db
conn = sqlite3.connect('bottle.db')
bottle_limpio.to_sql('bottle', conn, if_exists='replace', index=False)

# 4. Verificar el número final de columnas guardadas
cols_guardadas = pd.read_sql_query("PRAGMA table_info(bottle);", conn)['name'].tolist()
conn.close()

print(f"Columnas originales en bottle: {bottle.shape[1]}")
print(f"Columnas guardadas en 'bottle.db': {len(cols_guardadas)}")

Columnas originales en bottle: 74
Columnas guardadas en 'bottle.db': 26


Ahora si, qué columnas finales nos quedaron en Bottle y en Cast

In [16]:
import sqlite3
import pandas as pd
import re

# 1. Leer el archivo Markdown y construir el diccionario de descripciones
ruta_markdown = "/workspaces/proyecto-final-data-science--ECS/documentos/diccionario_variables_calcofi.md"

diccionario_md = {}
with open(ruta_markdown, "r", encoding="utf-8") as f:
    for linea in f:
        # Busca patrones tipo: - **NombreVariable**: Descripción  o  | NombreVariable | Descripción |
        match = re.search(r"[*|`]?([A-Za-z0-9_]+)[*|`]?\s*[:|-|\|]\s*(.*)", linea)
        if match:
            var_nombre = match.group(1).strip()
            var_desc = match.group(2).strip().strip("|").strip()
            if var_nombre and var_desc:
                diccionario_md[var_nombre] = var_desc

# 2. Obtener las variables actuales de bottle.db
conn_b = sqlite3.connect("bottle.db")
vars_bottle = pd.read_sql_query("PRAGMA table_info(bottle);", conn_b)['name'].tolist()
conn_b.close()



# --- IMPRIMIR BOTTLE CON DICCIONARIO MD ---
print(f" VARIABLES EN 'bottle.db' ({len(vars_bottle)} en total):")
print("=" * 80)
for i, var in enumerate(vars_bottle, 1):
    desc = diccionario_md.get(var, "Descripción no encontrada en el .md")
    print(f"{i:2d}. {var:<12} -> {desc}")

print("\n" + "=" * 80 + "\n")



 VARIABLES EN 'bottle.db' (26 en total):
 1. Cst_Cnt      -> ID de muestreo
 2. Btl_Cnt      -> ID de botella/muestra
 3. Sta_ID       -> ID de estación
 4. Depth_ID     -> ID de profundidad/muestra
 5. Depthm       -> Profundidad
 6. T_degC       -> Temperatura del agua
 7. Salnty       -> Salinidad
 8. O2ml_L       -> Oxígeno disuelto
 9. STheta       -> Densidad potencial (sigma-theta)
10. RecInd       -> Indicador de registro
11. T_prec       -> Precisión de temperatura
12. S_prec       -> Precisión de salinidad
13. NH3q         -> Calidad de amonio
14. C14A1q       -> Calidad C14 réplica 1
15. C14A2q       -> Calidad C14 réplica 2
16. DarkAq       -> Calidad botella oscura
17. MeanAq       -> Calidad de asimilación media
18. R_Depth      -> Profundidad reportada
19. R_TEMP       -> Temperatura reportada
20. R_POTEMP     -> Temperatura potencial reportada
21. R_SALINITY   -> Salinidad reportada
22. R_SIGMA      -> Densidad potencial reportada
23. R_SVA        -> Anomalía de volumen

In [15]:
# 3. Obtener las variables actuales de cast.db
conn_c = sqlite3.connect("cast.db")
vars_cast = pd.read_sql_query("PRAGMA table_info(cast);", conn_c)['name'].tolist()
conn_c.close()

# --- IMPRIMIR CAST CON DICCIONARIO MD ---
print(f" VARIABLES EN 'cast.db' ({len(vars_cast)} en total):")
print("=" * 80)
for i, var in enumerate(vars_cast, 1):
    desc = diccionario_md.get(var, "Descripción no encontrada en el .md")
    print(f"{i:2d}. {var:<12} -> {desc}")

 VARIABLES EN 'cast.db' (61 en total):
 1. Cst_Cnt      -> ID de muestreo
 2. Cruise_ID    -> ID de campaña
 3. Cruise       -> Campaña
 4. Cruz_Sta     -> Campaña + estación
 5. DbSta_ID     -> ID de estación de base
 6. Cast_ID      -> ID del lance/muestreo
 7. Sta_ID       -> ID de estación
 8. Quarter      -> Trimestre
 9. Sta_Code     -> Código de estación
10. Distance     -> Distancia a costa
11. Date         -> Fecha
12. Year         -> Año
13. Month        -> Mes
14. Julian_Date  -> Fecha juliana OA
15. Julian_Day   -> Día juliano
16. Time         -> Hora
17. Lat_Dec      -> Latitud decimal
18. Lat_Deg      -> Grados de latitud
19. Lat_Min      -> Minutos de latitud
20. Lat_Hem      -> Hemisferio de latitud
21. Lon_Dec      -> Longitud decimal
22. Lon_Deg      -> Grados de longitud
23. Lon_Min      -> Minutos de longitud
24. Lon_Hem      -> Hemisferio de longitud
25. Rpt_Line     -> Línea reportada
26. St_Line      -> Línea estándar
27. Ac_Line      -> Línea calculada
28. Rpt_S

In [21]:
# Variables clave seleccionadas para el análisis EDA
cols_usar_bottle = [
    'Cst_Cnt',   # Clave de unión
    'Btl_Cnt',   # ID muestra
    'Depthm',    # Profundidad (m)
    'T_degC',    # Temperatura (°C)
    'O2ml_L',    # Oxígeno disuelto (mL/L)
]

cols_usar_cast = [
    'Cst_Cnt',   # Clave de unión
    'Cruise',    # ID del crucero
    'Date',      # Fecha
    'Year',      # Año
    'Month',     # Mes
    'Lat_Dec',   # Latitud decimal
    'Lon_Dec'    # Longitud decimal
]

In [ ]:
import sqlite3
import pandas as pd

# Conectar y leer
conn_b = sqlite3.connect('bottle.db')
df_bottle = pd.read_sql_query("SELECT * FROM bottle;", conn_b)

# Filtrar dejando solo las seleccionadas
df_bottle_recortado = df_bottle[cols_usar_bottle]

# Sobrescribir la base de datos con if_exists='replace'
df_bottle_recortado.to_sql('bottle', conn_b, if_exists='replace', index=False)

# Verificar
vars_finales_b = pd.read_sql_query("PRAGMA table_info(bottle);", conn_b)['name'].tolist()
conn_b.close()

print(f"✅ 'bottle.db' actualizada. Variables conservadas ({len(vars_finales_b)}): {vars_finales_b}")

: 